### ISSUE TITLE
Purgeable resources

### ISSUE THREAD

**Issue by litherum** [gpuweb/gpuweb#647]
Not all resources are used all the time by a 3D application. In particular, applications often have caches of resources which are rarely used and low priority. Many times, the contents of these resources can be recreated on-demand.

This represents a classic space-time tradeoff. If there is plenty of memory, it makes sense for an application to keep these resources around so they can be used without having to recreate them (spending memory for time's benefit). However, if available memory gets low, these resources are good candidates to destroy because their contents can be recreated, and they are unlikely to be needed immediately (spending time for memory's benefit).

Because the web author doesn't know the exact machine their code is running on, they often are unable to make the right space-time tradeoff. Instead, the OS or browser has a much better knowledge of the system resources than the website. It would be valuable of the OS or browser could help the application make this tradeoff.

Purgeable resources are one way of doing this. Each resource could have something like:

```WebIDL
```

**Comment by Kangz** [comment]
I think this would be very useful but I'm not sure how it would be implementable on OS primitives for purgeability in multiprocess browser architectures.

**Comment by litherum** [comment]
Perhaps setNonPurgeable() could return a promise

**Comment by kvark** [comment]
Isn't an investigation missing here? For example, how are you planning to implement that on macOS?

**Comment by litherum** [comment]
This feature can be implemented by the browser or the OS. It can be implemented by the browser deleting the resource when the memory pressure handler is run (or when any other condition is hit). I think this is equally applicable to all platforms/APIs.

**Comment by kvark** [comment]
Hmm, sorry, I'm still not seeing the whole picture there. It would help to provide a few concrete entry points that would back that up on OS level. Take a recent #551 for example, where for each API there is a suggested path to implement a feature.

Looking at Metal, I see [useResource](https://developer.apple.com/documentation/metal/mtlrendercommandencoder/3043404-useresource?language=objc), which needs to be called, which makes me assume the driver/OS are always trying to make some GPU memory non-resident, and I don't see where this can be controlled.

Deleting a resource sounds like a scary business, since we don't always even know the contents of it on CPU, and we would have to re-create all the bind groups and views of that resource whenever it's resurrected.

**Comment by Kangz** [comment]
> Deleting a resource sounds like a scary business, since we don't always even know the contents of it on CPU, and we would have to re-create all the bind groups and views of that resource whenever it's resurrected.

Good point, maybe another way to do this, would be to have "discardable" resources that you have to lock to use, but can be destroyed (as if `destroy` was called) at any point when unlocked.

**Comment by litherum** [comment]
@kvark i think what your looking for is https://developer.apple.com/documentation/metal/mtlresource/1515898-setpurgeablestate

**Comment by RafaelCintron** [comment]
D3D has a similar concept to this called Offer/Reclaim. 

If you, the application developer, have a texture that you have the means to re-create easily, you can offer it to the Windows by calling [OfferResources](https://docs.microsoft.com/en-us/windows/win32/api/dxgi1_5/nf-dxgi1_5-idxgidevice4-offerresources1).  If the OS is running low on memory (GPU or CPU), it can elect to take your offered memory and give it to another process.  Taking an offered resource is cheaper than a normal resource because OS does not have to readback the resource from GPU memory or page the resource out to disk. 

If you, the application developer, wish to use memory you have offered, you must call [ReclaimResources](https://docs.microsoft.com/en-us/windows/win32/api/dxgi1_5/nf-dxgi1_5-idxgidevice4-reclaimresources1) before you can use the resource.  If the OS decided to take your resource, `ReclaimResources` will return an error and you'll need to re-create the resource.  If you're the only app running on the system and there was no memory pressure after all, `ReclaimResources` will return S_OK and you're free to continue to use the resource.  

We used Offer/Reclaim extensively in IE and old-Edge to offer all texture caches and image textures to the system. We kept around the compressed versions of images for re-decoding. Texture caches, we re-created if we found they were taken by the system.  

Unfortunately, Offer/Reclaim is not available on D3D12, but we could include a version of it in WebGPU and implement it in response to OS memory pressure notifications as Myles suggests.  The WebGPU implementation would simply destroy offered resources.  

While I do not think Offer/Reclaim (or whatever we call it) meets the bar for WebGPUCore, it would be good to consider for v2.  The alternative, in the meantime, is to forcibly lose WebGPU contexts.

**Comment by kainino0x** [comment]
I am in support of adding something like this eventually in WebGPU.

I think there are two possible models:

- (the one proposed above) If the system has purged a resource, then you get back the same resource but with its contents discarded. It would be similar to an object going from "destroyed" back to "not destroyed". (nit: I think setNonPurgeable would have to return a Promise.)
- If the system has purged a resource, then you have to create a new resource, create new bind groups from the resource, etc.

The former might might require implementations to automatically re-create backend objects for bind groups, command buffers, and command/pass encoders. That seems not ideal.

The latter could look like (pseudocode):

```WebIDL
interface GPUResourceBase {
    void setPurgeable();
    Promise<a new resource, or "original resource still ok"> setNonPurgeable();
}
```

**Comment by Kangz** [comment]
Agreed that "un-destroying" resources would be unfortunate because of all the state that would need to be recreated deep in `GPUCommandBuffers`, `GPUBindGroups` etc.

It seems that in a lot of cases an application might not be able to know even a frame in advance whether it will need a purgeable resource or not. Since the promise makes the API very difficult to integrate and use, here's some ideas of how purgeability/discardability could be implemented with a synchronous API.

For browsers in multi-process mode, it's not possible to use Metal's setPurgeable or D3D12's OfferResources, because we can't synchronously reclaim the resource on the renderer/content/Web process side. This means that instead the discarding of resources is done by the browser based on memory budgets or memory pressure events.

Two possible implementations when the browser handles purgeability itself:

 - Have an atomic flag in shmem for each purgeable resource that's one of (not-purgeable, purgeable, purged). When the browser decides to purge the resource, it first atomic compare swap `purgeable -> purged` and setNonPurgeable does an atomic compare swap `purgeable -> non-purgeable`.
 - When the GPU process thinks it needs to purge a resource, it sends a message to the renderer process and waits for its answer that the resource is still purgeable before purging (if the resource is still purgeable, the renderer process would tag it as purged when it receives the message).

**Comment by kvark** [comment]
It looks like D3D12's `OfferResources` requires the user to recreate resources. It doesn't seem feasible for us, since we don't want to be re-creating all the dependent resources (bind groups, views, bundles) behind the user's back.

@litherum how does the reclamation happen in Metal? Is it a matter of switching the purgeable state? calling `useResource`? or recreating the resource from the ground up?

**Comment by jdashg** [comment]
OfferResources is a cool design. It's like allowing parts of the device/context to get something-like-context-loss and need to be recreated. Having to recreate dependent resources is a pain, but it's better than having the whole context lost!

**Comment by RafaelCintron** [comment]
To be clear "OfferResources" is NOT in D3D12, at least not yet. 

In WebGPU, we could simulate it by tracking which resources have been least frequently used and "taking" ones the developer has offered in response to budget change notifications.  My thought was to have the developer be responsible for re-creating taken resources as well as dependent resources.  From their perspective, a taken resource is as good as destroyed, and should be treated as such from a validation perspective. 

The asynchronous nature of WebGPU implementations does make implementing a synchronous "Reclaim" APIs a challenge.

**Comment by litherum** [comment]
@kvark

> how does the reclamation happen in Metal? Is it a matter of switching the purgeable state? calling useResource? or recreating the resource from the ground up?

Calling `setPurgeableState(.volatile)` can't delete the resource itself; it can just make the resource internally-null. When you want to reclaim the resource, you just call `setPurgeableState(.nonVolatile)` and it will return either `.empty` or `.nonVolatile`. If it returns `.nonVolatile` then the resource is intact; you can use it as-if you had never done any of this volatility business at all. If it returns `.empty` then you aren't allowed to use it (and the developer has to make a new one).
